### DISCUSSION TITLE
Using excel data in nicegui get toggle button's thumb (indeterminate state) not proper

### DISCUSSION THREAD

**Discussion by Shivar90** [zauberzeug/nicegui#2472] [Q&A]
### Question

Hi Nicegui community I am using excel 
[ColumnsData.xlsx](https://github.com/zauberzeug/nicegui/files/14100613/ColumnsData.xlsx)
to create the table but when  table is create using below mentioned  code, toggle button is not looking proper always toggle button thumb (slider in between) I think indeterminate state  .Refer snap "Table Data" to see toggle button issue. 
But when create table using hardcoded values in columns and rows I am getting toggle button correctly . This issue I observed mainly when create table using excel and add toggle button per row based on Action column condition. (as mentioned above) 
Please explain how to resolve this toggle button thumb issue and give your opinion how to handle this issue.
<img width="918" alt="Table Data" src="https://github.com/zauberzeug/nicegui/assets/36196330/03a6573e-8b47-45d4-a804-787ce1e1d129">

```Python
from nicegui import ui, events
import json
import requests
import pandas as pd

# Define columns
columns = [
    {'name': 'action', 'label': 'Action', 'field': 'action', 'required': True, 'align': 'left'},
    {'name': 'Group', 'label': 'Group', 'field': 'Group', 'required': True, 'align': 'left'},
    {'name': 'Task', 'label': 'Task', 'field': 'Task', 'required': True, 'align': 'left'},
    {'name': 'Tags', 'label': 'Tags', 'field': 'Tags', 'required': True, 'align': 'left'},
    {'name': 'email', 'label': 'email', 'field': 'email', 'required': True, 'align': 'left'},
]

rows = [

]
# Read data from Excel sheet
df = pd.read_excel("ColumnsData.xlsx")
print("df data", df)

def handle_action_event(e: events.GenericEventArguments) -> None:
    print(f"{e.sender}")  # this will be the table object
    print(f"{e.args}")  # this will be the args defined in the emit call
    ui.notify(f'{e.args[1]} wants action {e.args[2]}')
    for row in rows:
        if row['id'] == e.args['id']:
            row['name'] = e.args['name']
    ui.notify(f'Table.rows is now: {table.rows}')

# Create the table with defined columns, add df data to table
table = ui.table(columns=columns, rows=df.to_dict('records'), row_key='name').classes('w-full')
Columns_Name = table.columns
print("Columns Name:", Columns_Name)
if any(column['label'] == 'Action' for column in table.columns):
    print(f"The table contains the 'Action' column.")
    table.add_slot('body-cell-action', '''
        <q-td key="action" :props="props">
            <q-toggle 
                :color="props.value ? 'red' : 'green'" 
                keep-color
                :label="props.value ? 'Get' : 'Leave'" 
                checked-icon="clear" 
                v-model="props.row.action" 
                v-model="check_value" 
                @update:model-value="() => $parent.$emit(
                    <!-- callback function -->
                    'action_event', <!-- should match string passed to `table.on` below -->
                    
                    <!-- arguments, can be whatever you want really -->
                    props.row,   <!-- this would pass the whole row of data -->
                    props.row.Group,   <!-- this would pass data of the Group column for this row -->
                    props.value ? 'Get' : 'Leave'  <!-- this will pass a string indicating the action based on the toggle switch state -->
                )"
            />           
        </q-td>
    ''')
else:
    print(f"The table does not contain the 'Action' column.")


ui.run(port=8080)
```

**Comment by falkoschindler** [comment]
Hi @Shivar90,
Can you, please, try to create a [minimal reproducible example](https://en.wikipedia.org/wiki/Minimal_reproducible_example)?

- Your code contains lots of comments, print statements, imports and edge cases seemingly unrelated to the problem.
- More importantly: I'm hesitating to download and open Excel files from the internet. Can you create a dataframe showing the same problematic behavior?

Long story short: Reducing the code (and data) to the bare minimum is the first step to the solution. Anything you can remove allows us to help more efficiently. Thanks!

**Reply by Shivar90** [reply]
Hi @falkoschindler Thanks for reply. I have created minimal reproducible example based on your comments. kept necessary thing to produce this error . Please check below code and Refer Snap observe issue of toggle button thumb (slider in between). As I mentioned above I observed this error when pass excel sheet data to Table not with hardcode data frames (columns and rows). Kept import and print statement related to this issue. 
[ColumnData1.xlsx](https://github.com/zauberzeug/nicegui/files/14163251/ColumnData1.xlsx)
Also import install openpyxl package (I am using version 3.1.2) for excel .
Please give any solution related to toggle button observed issue.
<img width="628" alt="Togglebutton issue in Table using excel" src="https://github.com/zauberzeug/nicegui/assets/36196330/ec75e189-6551-4aee-93ab-827284566c75">

MRE Code
```Python
from nicegui import ui
import pandas as pd

# Define columns
columns = [
    {'name': 'action', 'label': 'Action', 'field': 'action', 'required': True, 'align': 'left'},
    {'name': 'Group', 'label': 'Group', 'field': 'Group', 'required': True, 'align': 'left'},
]

rows = [

]
df = pd.read_excel("ColumnData1.xlsx")

# Create the table with defined columns, add df data of Excel to table
table = ui.table(columns=columns, rows=df.to_dict('records'), row_key='name').classes('w-full')
Columns_Name = table.columns
if any(column['label'] == 'Action' for column in table.columns):
    table.add_slot('body-cell-action', '''
        <q-td key="action" :props="props">
            <q-toggle :toggle-order="t":color="props.row.action ? 'green' : 'red'" v-model="props.row.action"/>
            <span v-if="props.row.action">Get</span>
            <span v-else>Leave</span>
        </q-td>  
    ''')
else:
    print(f"The table does not contain the 'Action' column.")

ui.run(port=8080)
```

**Reply by falkoschindler** [reply]
Hi @Shivar90, thanks for boiling it down!
As far as I can tell, there is no "Action" column in the Excel file. Therefore it makes sense for the toggle to remain in an "undefined" position.
![Screenshot 2024-02-05 at 11 20 53](https://github.com/zauberzeug/nicegui/assets/5767091/86cc1a94-5dc7-4fb6-9ff5-c5bd544196a4)

**Reply by Shivar90** [reply]
@falkoschindler Even I have added "Action" Column in Excel file ,added toggle button state as 't' /'true' /toggle-order='t' and 'f' ,'false'/toggle-order='false (tried different mentioned toggle value in excel sheet ) by referring "refer https://quasar.dev/vue-components/toggle#qtoggle-ap)" documentation. Executed same MRE code still get same error (even excel sheet and toggle button code in same folder ). Please check attached snap. 

 My original problem statement is divided in two parts
1. When table is created  using excel sheet, toggle button state is not proper "thumb (slider part)  is between -undefined position"
2. When I click toggle button toggle button status change to true and UI display "Get " and when clicked refresh (Press F5), toggle button state is changed to undefined position  & "Leave" toggle button's UI Status displayed -(toggle button's thumb in middle). Expected: After click toggle button should change to true state and even refresh page state remains same unless and until user click the toggle button .When clicked toggle button to false state and UI display "Leave " toggle button's UI Status display  and when clicked refresh (Press F5), toggle button state is changed to undefined position -(toggle button's thumb in middle) & Leave Status display. Expected: After click toggle button again  (if in true state) should change to false state and even refresh page state remains same unless and until user click the toggle button. Also toggle button state and UI status depend on Backend REST API value.

Please refer original Code for complete functionality .
Please check " handle_action_event" function and its usage in above code in this discussion. Please give any inputs and solution for this toggle button issue .  
[ColumnData1.xlsx](https://github.com/zauberzeug/nicegui/files/14168584/ColumnData1.xlsx)

<img width="664" alt="TogglebuttonissueinTable" src="https://github.com/zauberzeug/nicegui/assets/36196330/c0663d36-b9a5-4193-8679-3a7d5c126040">

**Reply by Shivar90** [reply]
Hi @falkoschindler I am able to resolve the toggle button thumb issue actually toggle button value is depend on REST API value. Suppose toggle _button_value is 'False' I set "props.row.value based on that. Thanks for help
